### ISSUE TITLE
Regular weight letters are rendered slightly taller than their bold variants

### ISSUE THREAD

**Issue by burodepeper** [chrissimpkins/Hack#95]
See https://github.com/atom/atom/issues/8620 for exact details.

**Comment by chrissimpkins** [comment]
1. Is this ttf or otf build? and what version?
2. Mind checking a set of regular glyphs immediately adjacent to bold glyphs in that editor (i.e. taking the gutter characters out of the equation)?

Similar issue reported in #32.  Unclear what is causing your issue at the moment.

**Comment by burodepeper** [comment]
v2_010-ttf\
I see that there is a 013 now, I'll give that a try tomorrow and check 2 with it.

On a side note, I'm on a mac, if I have the option, should I go for ttf or otf?

**Comment by chrissimpkins** [comment]
Not sure that it will fix this issue, but worth a try.  

> should I go for ttf or otf

You are trying to optimize cross-platform display for a syntax highlighter, correct?  If so, use the ttf files and let's work off of those.

**Comment by burodepeper** [comment]
Cross-platform is not necessarily my aim (but a nice plus), but I take from your answer that otf would be the better choice in itself?

**Comment by chrissimpkins** [comment]
Otf will work for screen use on Mac, not as well for screen on Windows and some Linux distros.

**Comment by burodepeper** [comment]
Not sure if this is what you wanted, but I've snapshotted several things with v013. The top and bottom greens should be the same, and all three variations with bold characters produce the error.

![schermafbeelding 2015-09-09 om 08 50 33](https://cloud.githubusercontent.com/assets/3703530/9755181/5069e00c-56d0-11e5-98a1-738e48af26b2.png)

![schermafbeelding 2015-09-09 om 08 52 43](https://cloud.githubusercontent.com/assets/3703530/9755191/5a34fcf2-56d0-11e5-8371-86fac70dc5a1.png)

![schermafbeelding 2015-09-09 om 08 52 33](https://cloud.githubusercontent.com/assets/3703530/9755187/568310b2-56d0-11e5-81f3-4d8f6c120db7.png)

![schermafbeelding 2015-09-09 om 08 52 17](https://cloud.githubusercontent.com/assets/3703530/9755185/53b578de-56d0-11e5-9442-2bbb64d6eedb.png)

**Comment by chrissimpkins** [comment]
Intend to start work on this over the  weekend.  Sorry for the delay. 



—
Chris

**Comment by burodepeper** [comment]
No problem at all.
Let me know if I can help in any way.

**Comment by chrissimpkins** [comment]
Are you seeing this at any text sizes other than 14px?

**Comment by burodepeper** [comment]
8px: **yes**
9px: **yes**
10px: no
11px: **yes**
12px: no
13px: no
14px: **yes**
15px: no
16px: no
17px: no
18px: no

**Comment by chrissimpkins** [comment]
that's helpful.  looking at the hinting today

**Comment by chrissimpkins** [comment]
I can confirm that the bold set is aligned to the baseline.  The baseline is shaded in the images below and I confirmed that the characters are positioned at 0 units (i.e. right on the baseline) where there are no bowls that overhang.  The latter is an intentional design issue to make these curved lower areas of the glyphs appear to align on the baseline.  If you don't provide this overhang, they appear to high relative to other glyphs.

Here are the bold upper case:
<img width="958" alt="caps" src="https://cloud.githubusercontent.com/assets/4249591/9905284/1de8a2d0-5c52-11e5-86d8-9bd8a8183383.png">

And the bold lower case:
<img width="950" alt="lower" src="https://cloud.githubusercontent.com/assets/4249591/9905293/23b20dd2-5c52-11e5-945b-4a9f5825a3fb.png">

Let's check one thing before I dig into the hints.  Do you mind setting the bold characters in your editor to **only** those that do not have an overhang past the baseline.  Use only glyphs like the `Ff`, `Hh`, or `Tt`.  I'd like to confirm that this isn't causing these to snap up by a pixel for some reason.

**Comment by weslem** [comment]
Hi, I've experienced this same issue. I've captured this on an up-to-date x86_64 ArchLinux box using the TTF fonts.

* fontconfig settings: DPI=161, lcdfilter=lcddefault, rgba=rgb.
* terminal: st, using a font setting of Hack:size=10

![bold_capture](https://cloud.githubusercontent.com/assets/994468/9916048/5a2c4a36-5c6f-11e5-9501-d4bcc91858cb.png)

Notice the cursor character renders correctly for some reason. I verified that the reverse-video cursor shows the correct alignment on all the affected characters. Similarly, bold-oblique characters appear correct:
![italic_capture](https://cloud.githubusercontent.com/assets/994468/9916050/5d21ced2-5c6f-11e5-908a-72151bc82ee3.png)

**Comment by chrissimpkins** [comment]
@weslem thank you very much.  I am assuming this is with FreeType?  Are you using any modifications to FreeType?

**Comment by weslem** [comment]
Correct, the captures above are from an unmodified (to my knowledge) freetype2. The st terminal is a tiny program (one C file), and the font code doesn't appear that complicated to me: http://git.suckless.org/st/tree/st.c
I'm not familiar enough with font rendering quickly to identify any problems, however.

I can reproduce similar output with other freetype2 terminals rather easily. Here are some examples from my system:

```shell
xterm -fa Hack:size=12 -bg black -fg white -e \
    'for char in A F f H h T t x . _; do echo -e "${char}\033[1m${char}\033[0m${char}"; done; read'
uxterm -fa Hack:size=12 -bg black -fg white -e \
    'for char in A F f H h T t x . _; do echo -e "${char}\033[1m${char}\033[0m${char}"; done; read'
st -e sh -c 'for char in A F f H h T t x . _; do echo -e "${char}\033[1m${char}\033[0m${char}"; done; read'
```

urxvt, which does not link to freetype2, seems to work correctly:

```shell
urxvt -fn xft:Hack:size=12 -bg black -fg white -e \
    sh -c 'for char in A F f H h T t x . _; do echo -e "${char}\033[1m${char}\033[0m${char}"; done; read'
```

**Comment by chrissimpkins** [comment]
So I checked the hints with the freetype grid viewer and this does not appear to be the source of the problem.  Here are the hinted freetype2 renders on the grid at 8pt and 18pt.  I looked at all text sizes from 6 - 32pt in both the regular and bold set.  The bold glyphs never get nudged off of the baseline.  

#### Regular
<img width="316" alt="lower-a-8" src="https://cloud.githubusercontent.com/assets/4249591/9919602/0d4464a6-5c9c-11e5-8be8-adc28c1402b3.png">

<img width="319" alt="lower-a-18" src="https://cloud.githubusercontent.com/assets/4249591/9919603/10699fac-5c9c-11e5-8436-4a345ed17ab9.png">

#### Bold
<img width="321" alt="lower-a-bold-8" src="https://cloud.githubusercontent.com/assets/4249591/9919607/145a4d32-5c9c-11e5-9cd8-9efbd6cb33d4.png">


<img width="318" alt="lower-a-bold-18" src="https://cloud.githubusercontent.com/assets/4249591/9919610/195dd330-5c9c-11e5-80df-dff71f46bd54.png">

Let me look at the vertical metrics tonight.  I may generate a few test builds with different vertical metrics if you are willing to try them.

**Comment by burodepeper** [comment]
I had to do a clean re-install of Atom this morning. I switched back to the default theme (without bold letters in this example) and removed any personal styling, and this was the result:

![menlo-vs-hack](https://cloud.githubusercontent.com/assets/3703530/9958062/13edc69c-5e06-11e5-9cfe-2a740eb2382e.jpg)
Left: `Menlo 14px`
Right: `Hack 14px`

The line-height is set at 1.5 (so should be 21px), but all lines in the Hack example are rendered as 22px high. The lines are grouped in absolute positioned sets of six. That's why the seventh line is properly aligned again.

So the left of the image is everything completely set to default, and in the right half I've only changed the typeface to Hack. This is _weird_.

**Comment by chrissimpkins** [comment]
Sorry for the delay.  I will try to have new builds with adjustments to the vertical metrics out this weekend.  I think that this is coming from those settings but it is not clear to me which at the moment.  We will need to play with some adjustments to see if we can figure this out.

**Comment by burodepeper** [comment]
No problem at all. I'm actually excited because (to me at least) it's such a weird issue.

**Comment by chrissimpkins** [comment]
New test build v2.014 available with new vertical metrics. I hope that this takes care of this issue.  Please let me know how it works.

Files are available in #111

**Comment by burodepeper** [comment]
Initial visual quick glance looks **great**. Will do some proper in depth testing later today.

**Comment by chrissimpkins** [comment]
Excellent!  Let me know.

**Comment by weslem** [comment]
I ran through the same tests I documented above. All the terminals look good to me now.

**Comment by chrissimpkins** [comment]
@weslem Excellent, thank you very much Matthew.  That is helpful.  I'm interested in both of your thoughts about the line spacing at this stage as well.  I am planning to make the new vertical metrics our default on the release builds and this is significantly tighter than the previous release.  We have another open issue report where we are discussing looser variants (and possibly tighter, but I think that this is where most who requested tighter line spacing want to be).

**Comment by burodepeper** [comment]
I only just noticed in my terminal that line-spacing had indeed changed. I am okay with this tightness. I have always been under the impression that I am to define the line-spacing in-app (ie, it hasn't changed anything for me in Atom for instance, since I have it set to 1.5 there), but you are talking about a default line-spacing when no other is defined?

About the other changes.
- Love the new tilde.
- I'd still prefer a 6-pointed centered asterisk, but this one suits the elegance of Hack better, in comparison to the more robust Menlo.
- I love how the dashes are horizontally really distinguishable, and I was wondering if it is by design that the underscore doesn't have this as well. At 14px I see that underscores are not completely connected, but I can't easily count them.

Besides my preferences, **kudos!**

**Comment by chrissimpkins** [comment]
Thanks for all of these comments.  

> but you are talking about a default line-spacing when no other is defined 

Correct, this will make the line spacing tighter than the previous version by default across all platforms and editors. Users will need to manually adjust line spacing to return to the appearance of the last version.  

> I'd still prefer a 6-pointed centered asterisk, but this one suits the elegance of Hack better, in comparison to the more robust Menlo.

I am considering some alternate builds (Input style) and we could consider this change in one of these builds.  The lowercase `i` and the zero seem to be polarizing glyphs as well.  You can define alternative glyph forms via OpenType but most text editors don't support this feature at this stage.

> I love how the dashes are horizontally really distinguishable, and I was wondering if it is by design that the underscore doesn't have this as well 

This is a *proposed* change for the next release and did result in a change in the vertical orientation, centering, width, and height of the underscore.  The width change is based upon a request in #97 .  I agree with you.  I like to see the spaces between the glyphs as well and did not want to go too far into the ligature range where it simply looks like a solid line.  Feel free to join us over there to discuss this further.  I can look at the hinting to see if we are losing a pixel or two between the glyphs at certain text sizes.  This might lead to widening the design or altering the hints (or both).

**Comment by burodepeper** [comment]
Cool, I'm coming over to #97.
Could you recommend software to look at the glyphs and stuff myself? I'm rather curious what it all looks like under the hoofd.

**Comment by chrissimpkins** [comment]
[FontForge](https://fontforge.github.io/en-US/) is excellent and libre/free.  It uses the X Window system on OS X and looks terrible on Retina displays which is why I don't use it.  It makes it really tough on the design side of things.  It's great for scripting and low resolution viewing there and may look fine on Windows / Linux.  I haven't tried it on these platforms.

I currently use [Glyphs](https://www.glyphsapp.com/) and [FontLab Studio](http://old.fontlab.com/font-editor/fontlab-studio/).  [RoboFont](http://doc.robofont.com/) is another popular commercial alternative.  

All of the above support scripting (generally in Python).   I'd be happy to discuss my impressions of the paid applications off repository if you are interested in my impressions before you sink money into software.

**Comment by weslem** [comment]
Overall, I like all the improvements I've seen. I've got the test build as my primary font today at work, and the new line spacing feels quite comfortable. I especially like the new tilde.

The changes don't quite address #104, however. The vertical line character really should not have any separation at all, and tmux splits still look wrong.

**Comment by chrissimpkins** [comment]
@weslem Thanks. I will take a look at this and see if we can address it in this release.  There are a number of vertical character-to-character alignment issues outstanding that required a constant set of vertical metrics to address (box glyphs, line glyphs, etc).

**Comment by JesusMtnez** [comment]
Is it fixed?

**Comment by chrissimpkins** [comment]
@JesusMtnez should be.  Mind giving the test build a try?  It's available in #111 .  Pushing this as a release over the weekend.

**Comment by JesusMtnez** [comment]
@chrissimpkins As you said, in 2_014 test build the issued is fixed. I'll be waiting for the stable release. Thank you!!

**Comment by chrissimpkins** [comment]
@JesusMtnez Great!  Hope to have it out in the next day or so.  Have a few small changes to add.

**Comment by chrissimpkins** [comment]
@JesusMtnez Any thoughts on the new line spacing?

**Comment by JesusMtnez** [comment]
@chrissimpkins I love it. I prefer it this way. Great job! :+1:

**Comment by chrissimpkins** [comment]
@JesusMtnez excellent, good to hear that.  That seems to be the consensus though there are some who wanted it even looser.  Once I have all of the vertically aligned glyphs in proper position, we will likely build some wider variants.

**Comment by chrissimpkins** [comment]
These changes are now included in release v2.015.  Thanks for the issue report David and thanks to all who helped with the testing of this issue here.
